File: problems/198/solution.go
package main
func rob(nums []int) int {
    if len(nums) == 0 {
        return 0
    }
    
    if len(nums) == 1 {
        return nums[0]
    }
    dp := make([]int, len(nums) + 1)
    dp[0] = 0
    dp[1] = nums[0]
    
    for i := 2; i <= len(nums); i++ {
        dp[i] = max(dp[i - 1], dp[i - 2] + nums[i - 1])
    }
    
    return dp[len(nums)]
}

func max(a, b int) int {
    if a > b { return a }
    return b
}

File: problems/198/solution.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        if (nums.size() == 0) {
            return 0;
        }
        
        if (nums.size() == 1) {
            return nums[0];
        }
        int i2 = 0, i1 = nums[0];
        
        for (int i = 2; i <= nums.size(); i++) {
            int tmp = i1;
            i1 = max(i1, i2 + nums[i - 1]);
            i2 = tmp;
        }
        
        return i1;
        
//         vector<int> dp(nums.size() + 1);
//         dp[0] = 0;
//         dp[1] = nums[0];
        
//         for (int i = 2; i <= nums.size(); i++) {
//             dp[i] = max(dp[i - 1], dp[i - 2] + nums[i - 1]);
//         }
        
//         return dp[nums.size()];
    }
};